crates/matrix-sdk-base/src/event_cache_store/wrapper.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{Arc, Mutex as StdMutex};
+
+use ruma::{time::SystemTime, MxcUri};
+use tokio::sync::Mutex as AsyncMutex;
+
+use super::{DynEventCacheStore, MediaRetentionPolicy, Result};
+use crate::media::MediaRequest;
+
+/// A wrapper around an [`EventCacheStore`] implementation to abstract common
+/// operations.
+///
+/// [`EventCacheStore`]: super::EventCacheStore
+#[derive(Debug, Clone)]
+pub struct EventCacheStoreWrapper {
```

Comment:
I understand why you've chosen this design.

I'm not comfortable with this though for several reasons, the major one being it introduces another type indirection for the `EventCacheStore` itself. To reach the `EventCacheStore`, the entry point is `EventCacheStoreWrapper`, then `DynEventCacheStore`, then `EventCacheStore`. If we include the `EventCacheStoreLock`, it introduces _another_ level of indirection.

Note that this is only for media management for the moment.

Possible solutions:

- Use standalone functions to be used by the multiple `EventCacheStore` implementations, liek we did for the cross-process lock and `MemoryStore`: https://github.com/matrix-org/matrix-rust-sdk/commit/94bd421a8db199fe6aedd7e12b6a1788059bda32
- Use a standalone type to manage media, like `MediaService`, which takes a `MediaRetentionPolicy`. The `EventCacheStore` would delegate the handling of media to `MediaService`, such as:

  ```rust
  pub async fn add_media_content(&self, request: &MediaRequest, content: Vec<u8>) -> Result<()> {
      self.media_service.add_media_content(&self, request).await
  }
  ```

That way we can test `MediaService` alone, it's a self-contained API, and it removes the need for another `EventCacheWrapper` type + type indirection.

Bonus: `MediaService` can be introduced inside its own commit, along with its tests. Another commit can use `MediaService` inside the `EventCacheStore` implementations.

What do you think?

Reply:
How does the `MediaService` access the data from the backend? Are we supposed to add a new trait for that?

Reply:
I'm sorry but this is still confusing to me. From what you are saying, my understanding is that an `EventCacheStore` (or `EventCacheStoreMedia` with what you just said) implementation would look like:

```rust
impl EventCacheStore for MemoryStore {
  async fn add_media_content(&self, request: &MediaRequest, content: Vec<u8>) -> Result<()> {
    self.media_service.add_media_content(&self, request, content).await
  }

  // other methods…
}
```

Which means the `MediaService` would look like this:

```rust
pub struct MediaService {
  media_retention_policy: Arc<StdMutex<Option<MediaRetentionPolicy>>>,
  media_cleanup_guard: Arc<AsyncMutex<()>>,
}

impl MediaService {
  // other methods…

  async fn add_media_content(&self, cache: &impl InnerMediaApi, request: &MediaRequest, content: Vec<u8>) -> Result<()> {
    let media_retention_policy = self.media_retention_policy();

    // maybe do some pre-checks with the policy…

    cache.add_media_content_inner(media_retention_policy, request, content).await
  }
}
```

We need a trait for a private API that is only accessed by the `MediaService`, right? The `MediaService` cannot call the methods from the `EventCacheStore` trait because they are calling it?

Reply:
What I propose is:

```rust
trait EventCacheStoreMedia {
    fn add_media_content(…);
    fn remove_media_content(…);
}

trait EventCacheStore: EventCacheStoreMedia { … }

impl EventCacheStoreMedia for MediaService { … }

impl EventCacheStore for MemoryStore {
    fn add_media_content(…) {
        self.media_service.add_media_content(…);
    }

    …
}
```

something roughly like that.

Reply:
But what API does the `MediaService` call in its implementation of `add_media_content`? Since we are storing/removing data from the backend, the `MediaService` needs a generic API to access it somehow.

Reply:
Yes. `MediaService<Store: EventCacheStore>`, something like this. Thoughts?